### PR TITLE
Get json paths for metric

### PR DESCRIPTION
**Pull request by kamikaz1k**
**In this PR:**
- Added a metrics/search_parameters?metric_name endpoint
- returns list of json paths available for a given metric name
- the json paths can then be used for aggregation metrics

### Empty data query 

`curl http://localhost:8000/metrics/search_parameters?metric_name=nothing`

```
// HTTP 200

{
    "data": {
        "parameter_names": []
    }
}
```

### Nested data query

*note: it selects the "newest" entry for a given metric name

`curl 'http://localhost:8000/metrics/search_parameters?metric_name=who'`

```
// HTTP 200

{
  "data": {
    "parameter_names": [
      "cloud",
      "cluster",
      "decimal_revision",
      "latency",
      "nested",
      "nested.level_1",
      "nested.level_1.level_2",
      "nested.level_1.level_2.level_3",
      "release",
      "revision"
    ]
  }
}
```

### Non-existent metric_name query

`curl 'http://localhost:8000/metrics/search_parameters?metric_name=doesnotexist'`

```
// HTTP 400

"metric name does not exist..."
```

### Entries in DB

```
metrix=# SELECT metric_name, data FROM metrics WHERE id > 5 LIMIT 10;
 metric_name |                                                                             data
-------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------
 who         | {"cloud":"gcp","cluster":"alpha-1","decimal_revision":9.0,"latency":3,"release":"1911.1.0","revision":9}
 dat         | {"cloud":"gcp","cluster":"alpha-1","decimal_revision":9.0,"latency":5,"release":"1911.1.0","revision":9}
 who         | {"cloud":"gcp","cluster":"alpha-1","decimal_revision":9.0,"latency":9,"nested":{"level_1":{"level_2":{"level_3":"done"}}},"release":"1911.1.0","revision":9}
 nothing     | {}
```

### CODE REVIEW

File: src/routes/metrics.rs [hunk 1]
```diff
@@ -6,6 +6,7 @@ use diesel::sql_query;
 use rocket::http::RawStr;
 use rocket::response::status::BadRequest;
 use rocket_contrib::json::Json;
+// use serde_json::Value;
```
Comment (woodgern) [inline src/routes/metrics.rs]: :rotating_light:

File: src/routes/metrics.rs [hunk 2]
```diff
@@ -57,6 +58,81 @@ pub fn query_metric_route(
...
+    }
+
+    let query_string = format!("SELECT * FROM metrics WHERE metric_name = '{}' ORDER BY id DESC LIMIT 1", parsed_metric_name);
+    println!("### query_string: {}", query_string);
+    let query_result = sql_query(query_string)
+        .load::<Metric>(&db_conn)
+        .expect("Error loading metrics");
+
+    if query_result.len() == 0 {
+        return Err(BadRequest(Some("metric name does not exist...".to_string())));
```
Comment (woodgern) [inline src/routes/metrics.rs:82]: Maybe note that this should be a 404 later

File: src/routes/metrics.rs [hunk 3]
```diff
@@ -57,6 +58,81 @@ pub fn query_metric_route(
...
+fn get_paths_from_json(data: &serde_json::Value) -> Vec<String> {
+    let mut output: Vec<Vec<String>> = vec![];
+    let current_path = vec![];
+    deep_keys(&data, current_path, &mut output);
+
+    return output.into_iter().map(|keys| keys.join(".")).collect();
+}
+
+// https://stackoverflow.com/a/57275829
+fn deep_keys(value: &serde_json::Value, current_path: Vec<String>, output: &mut Vec<Vec<String>>) {
```
Comment (woodgern) [inline src/routes/metrics.rs:109]: R E C U R S I O N

File: src/routes/metrics.rs (hunk 2 above)
Comment (kamikaz1k) [reply src/routes/metrics.rs:82]: yea, I didn't wanna think about it

**Pull request merged**